---
title: "Introduction to the R Language Syntax and the Command Line (Terminal)"
author: "Gepoliano Chaves, Ph. D."
date: "August 28th, 2023"
output:
  # pdf_document: default
  # output: slidy_presentation
  revealjs::revealjs_presentation
  # html_document:
  #   df_print: paged
---
# 1) Basics of R

# 1.1) Basic Steps

* R can be accessed by clicking an icon or entering the command "R" at the system command line, also refered to as Terminal

* This produces a console window or causes R to start up as an interactive program at the current terminal window

* R works fundamentally by a question-and-answer model: enter a line with a command and press Enter

* Then the program does something relevant

# 1.1) Basic Steps (continued)

* In this course, we will use RStudio and a file format called R Markdown, which allows us to code using not only R, but bash and Python, two other languages we may refer to as the course progresses

# 1.2) Library ISwR

* This notebook used the book Introductory Statistics with R, by Peter Dalgaard as a reference

* For the book, R library ISwR (Introductory Statistics with R) can be freely downloaded

* All examples in the book used as reference should run provided that ISwR library is not only installed by also loaded into the current search path

# 1.2) Library ISwR (continued)

* Library can be installed and loaded by typing the following command into an R chunk

```{r, eval = F}
install.packages("ISwR")
library("ISwR")
```

# 1.3) Plot Random Numbers

* For a first impression of what R can do, let´s try plotting a graph

* Need to insert the R chunk and use the plot function

# 1.3) Plot Random Numbers (continued)

```{r}
plot(rnorm(1000))
```

# 1.4) A Potent Calculator

* R can process simple and complex arithmetic expression and produce a result for the user

```{r}
2 + 2
```

* R can also be used to do other standard calculations. Here is how to calculate e to the power of -2

```{r}
exp(-2)
```

# 1.4) A Potent Calculator (continued)

* Other than the R chunks, these calculations can be made using the RStudio Console

* In RStusio, when using the R Markdown format (.Rmd), we can insert R, Bash (terminal) and Python chunks

* Proceed to illustrate that in the RStudio IDE

# 1.5) Assignments

* Assignments are made based on the necessity to store the results of calculations and use these results in downstream processing steps in an entire algorithm or pipeline

* Like other languages, R has symbolic **variables**: names that can be used to represent values

* To assign the value 2 to variable x, we can enter the following command

```{r}
x <- 2
```

* The character <- is called the assignment operator

# 1.5) Assignments (continued)

* There is no immediate visible result, but from now on, x has the value 2 and can be used in subsequent arithmetic operations

* We can "ask" R which value is stores in the x variable

```{r}
x
```


# 1.6) Operations after assignment of variable

* Below, our variable x, is used to perform other calculations

```{r}
x
```

* Addition

```{r}
x + x
```

# 1.6) Operations after assignment of variable (continued)

* Multiplication

```{r}
5*x
```

* Potentiation

```{r}
x^3 
```

# 1.7) Vectorized Arithmetic

* It is not useful to use one number at a time to run statistics

* One strenght of R is that it can handle entire data vectors as single objects

* A data vector is an array of numbers and a vector variable can be constructed like this

```{r}
weight <- c(60, 72, 57, 90, 95, 72)

## To look at the vector variable, just type its name again
weight
```

# 1.7) Vectorized Arithmetic (continued)

* You can do calculations with vectors just like ordinary numbers, so long as they have the same length

* One exception to this rule that we will see will be when we use the mean of weigths of persons (represented by xbar)

* In that case, the mean will be one single number, which will be subtracted from each sample value

# 1.7) Vectorized Arithmetic (continued)

* Suppose the weight vector indicates the weight of men in kilograms

* One simple formula to indicate whether a person is obese or not, is the body mass index (BMI)

* BMI is calculated by dividing the person´s weight by the square of their height, in meters

# 1.7) Vectorized Arithmetic (continued)

* Therefore, in R, we need to have a vector with the height values to calculate the bmi vector, containing the body-mass index for the individuals indicated in the weight vector

```{r}
height <- c(1.75, 1.80, 1.65, 1.90, 1.74, 1.91)
bmi <- weight/height^2
bmi
```

# 1.8) Calculate the Mean of the variable weight

* The mean is calculated by the sum of the observations divided by the total number of observations

 $\overline{x}$ = $\sum x_i$ /n
 
# 1.8) Calculate the Mean of the variable weight (continued)

* Let´s calculate the mean of the variable weight

```{r}
sum(weight)

sum(weight)/length(weight)
```

# 1.9) Calculate the Standard-Deviation of the variable weight

* Standard-deviation can be calculated with the following equation

$$ SD = \sqrt{(\sum (x_i - \overline{x})^2)/(n-1)}$$

# 1.9) Calculate the Standard-Deviation of the variable weight (continued)

* xbar, the mean of variable weight, can be calculated using the sum and length of variable weight

```{r}
xbar <- sum(weight)/length(weight)
xbar
```

# 1.9) Calculate the Standard-Deviation of the variable weight (continued)

* Now we can calculate the difference of each replicate in the weight variable and the mean of the weight variable, one by one

```{r}
weight - xbar
```

# 1.9) Calculate the Standard-Deviation of the variable weight (continued)

* Notice how R uses xbar, which has length one, to calculate the new weight - xbar data vector

* xbar is recycled and subracted from every element in the weight variable (weight data vector)

# 1.10) Calculation of the Standard Deviation

* Calculate the squared deviations

```{r}
(weight - xbar)^2
```

* Calculate the sum of squared deviations

```{r}
sum((weight - xbar)^2)
```

#

Calculate the standard deviation

```{r}
sqrt(
  sum(
    (weight - xbar)^2/
      (length(weight)-1)
)
)
```

# 1.11) Standard Statistical Procedures

* It is a standard medical practice to access whether a person is obese or not using validated scientific criteria

* As a simple procedure to show this concept, let's assume that an individual with a normal weight should have a BMI in the range 20-25

* We want to know if our data deviates from the normal range of BMI

* In R, this can be done using a statistical test called t-test

#

* You do not need to understand what a t-test is, just remember that is is used to evaluate the distribution of sample values compared to the normal distribution

* You can use a one-sample t-test to assess whether the six persons' BMI can be assumed to have mean 22.5 given that they come from a normal distribution

* You can do that using the function t.test

# 1.11) Standard Statistical Procedures: t-test

```{r}
t.test(
  bmi,
  mu = 22.5
)
```

# 1.12) Plot Graphics

* Let's now plot a scatterplot of the height and weight of individuals

#

```{r}
plot(height, weight)
```

# 1.12) Plot Graphics Modifying the plotting character

* You will frequently want to modify drawing of your graphs in various ways

* One way is usig the parameter "plotting character", pch

#

```{r}
plot(height, weight, pch =2)
```

# 1.13) Plot an expected Line for weights at BMI of 22.5

* The term for weight at BMI of 22.5 is:

```{r, eval=F}
22.5*(height)^2
```

* This term can be used to compute the line of expected weights at a BMI of 22.5 using the lines() function

# 1.13) Plot an expected Line for weights at BMI of 22.5 (continued)

* As a note, we can’t use the lines() function without first creating a plot in R.

```{r}
plot(height, weight, pch =2)
hh <- c(1.65, 1.70, 1.75, 1.80, 1.85, 1.90)
lines(hh, 22.5 * (hh)^2)
```


# 1.14) Vectors

* The weight and height vectors are called numeric vectors

* Besides numeric vectors, there are numeric and character vectors

# 2) Using other libraries

## 2.1) The iris data-frame

* Now we have a better understanding of what R can give us, let us use another library more commonly used datasets

* At times, it is possible that you will need to figure out different ways to install a library to use it

```{r}
library(datasets)
```

#

* In the next chunk, we access the iris data, and look at a summary of the dataset

```{r}
data(iris)
summary(iris)
```

# 2.1) The iris data-frame

* Another form to look at the iris data-frame is typing its name

```{r}
iris
```

# 3.1) More Visualization of the iris dataset

 We need to cover these basic R functions:
 
* dim()
* head()
* View()
* class()
* str()

#

```{r, eval = F}
dim(iris)
head(iris)
View(iris)
class(iris)
str(iris)
```

# 3.2) Visualize iris dataset with basic R functions

* head

```{r}
head(iris, n = 10)
```

#

* number of columns and rows

```{r}
dim(iris)
```

#

* the class() function

```{r}
class(iris)
```

# 4) More Visualization

# 4.1) Scatterplot

```{r, eval=T}
plot(data=iris, iris$Sepal.Length, iris$Sepal.Width) ## R will complain about this command

plot(iris$Sepal.Length, iris$Sepal.Width) ## According to the error, this was plotted using a different synthax
```

# 4.1) Scatterplot

Plot without warning and message

```{r, eval=T, warning=F, message=F}
plot(data=iris, iris$Sepal.Length, iris$Sepal.Width) ## R will complain about this command

plot(iris$Sepal.Length, iris$Sepal.Width) ## According to the error, this was plotted using a different synthax
```

# 4.2) Boxplot

```{r, eval=T}
boxplot(data=iris, iris$Sepal.Length, iris$Sepal.Width)
```

# 5) Data Visualization with specialized libraries

* In R, there are packages designed with the purpose of making good-looking graphics. This is the case with the ggplot2. The chunk below installs the ggplot2 library, loads the library into the R environment and then plots the data present in the iris data-frame.

* You can uncomment the installation line if you need to install it

# 5.1) Ggplot2

```{r, eval = T}
#install.packages("ggplot2")
library(ggplot2)
ggplot(data=iris, aes(x=Sepal.Length, y=Sepal.Width, color=Species)) + geom_point(size=4)
```

In our activities to visualize human genomic data, we will use a library called qqman, to visualize the biological association, through a plot known as the Manhattan plot.

# 5.2) Visualizing GWAS

* The simplest definition of a GWAS is the statistical or significant association between a phenotype (trait) and a genotype. This association can also be called *biological association*.

* Information about association of SNPs with Huntington's Disease can be found at the Chaves 2019 Huntington's disease paper

# 5.3.1) Package installation

* We need to install and load package qqman

```{r, eval=T}
## install.packages("qqman")
library(qqman)
```

# 5.3.1) Package installation

* After, we load data-frame to be visualized

* Exact location of text file in your system needs to be determined

```{r, eval=T}
GWAS_TABLE <- read.table("../data/chr4.txt", 
                         sep = " ",
                         header = T)
```


# 5.3.2) Get information about object with head()

* Use head() function to inspect the data-frame just loaded

```{r, eval=T}
head(GWAS_TABLE, n = 10)
```

# 5.3.3) Plotting GWAS data-frame

* In this section we use functions plot() and boxplot() to visualize data-frame

* In the y access, we see genomic coordinates and in the x access, p-valoues of the biological association

* Note that depending on the synthax used for plotting, R may complain

#

* Comments not turned off

```{r, eval=T}
plot(data=GWAS_TABLE, GWAS_TABLE$BP, GWAS_TABLE$P) ## R complains
```

# 

* Turn off comments

```{r, eval=T, warning=F, message=F}
plot(data=GWAS_TABLE, GWAS_TABLE$BP, GWAS_TABLE$P) ## R complains
```

# 5.3.4) Plotting GWAS data-frame

```{r, eval=T}
plot(GWAS_TABLE$BP, GWAS_TABLE$P) ## R does not complain
```

# 5.4) Boxplot

* Boxplot

```{r, eval=T}
boxplot(data=GWAS_TABLE, GWAS_TABLE$BP, GWAS_TABLE$P)
```

# 5.5) Boxplot

* Compare the chromosomal coordinates and the values of the p-values in the boxplot above

* The chunk below allows one to ommit NA values in data-frame

```{r, eval=T}
GWAS_TABLE_Ommit <- na.omit(GWAS_TABLE)
```

#

A seguir, criamos uma variável que armazena as posições das SNPs a serem realçadas em verde no Manhattan plot. Estas SNPs são mutações biológica e estatisticamente associadas à Doença de Huntington no gene de uma sortilina, localizada em proximidade física ao gene da proteína huntingtina mutada, a qual é a causadora medeliana (segue as leis de segregação genética de Mendel) da Doença de Huntington.

# 5.6) Create vector to highlight genome coordinates in Manhattan plot

```{r, eval=T}
SNP_HIGHLIGHT <- c("chr4:3043512","chr4:3043513","chr4:3048207","chr4:3224216",
                   "chr4:3231772","chr4:3233844","chr4:3235081","chr4:3235084",
                   "chr4:3236881","chr4:3236883","chr4:3241845","chr4:3243804",
                   "chr4:3263138","chr4:3265130","chr4:3265710","chr4:3314646",
                   "chr4:3380088","chr4:3409359","chr4:3411110","chr4:3415336",
                   "chr4:3415378","chr4:3438643","chr4:3446091","chr4:3449886",
                   "chr4:3473066","chr4:3476809","chr4:3480439","chr4:3487151",
                   "chr4:3496058","chr4:3496110","chr4:3506933","chr4:3508752",
                   "chr4:3510957","chr4:3512690","chr4:3517746","chr4:3518190",
                   "chr4:3529671","chr4:3532327","chr4:3533066","chr4:3746133",
                   "chr4:3747842","chr4:3748134","chr4:3765305","chr4:3765336",
                   "chr4:3944253","chr4:3944752","chr4:3944888","chr4:3946166",
                   "chr4:3946175","chr4:3969218","chr4:4051294","chr4:4076788",
                   "chr4:4103104","chr4:4103105","chr4:4109198","chr4:4109210",
                   "chr4:4240627","chr4:4242705","chr4:4243668","chr4:4245210",
                   "chr4:4245510","chr4:4245513","chr4:4245591","chr4:4245926",
                   "chr4:4245929","chr4:4246109","chr4:4246433","chr4:4246453",
                   "chr4:4246457","chr4:4246497","chr4:4249414","chr4:4249415",
                   "chr4:4249484","chr4:4271623","chr4:4275306","chr4:4304749",
                   "chr4:4318931","chr4:4318970","chr4:4319564","chr4:4319728",
                   "chr4:4319750","chr4:4322078","chr4:4709657","chr4:4732282",
                   "chr4:4789635","chr4:4822960","chr4:4824890","chr4:4825092",
                   "chr4:4825180","chr4:4865316","chr4:4865321","chr4:5018702",
                   "chr4:5812778","chr4:5814082","chr4:5833660","chr4:5833899",
                   "chr4:5835541","chr4:5851205","chr4:5862752","chr4:5862938",
                   "chr4:5862943","chr4:5901873","chr4:5905499","chr4:5906287",
                   "chr4:6018891","chr4:6019046","chr4:6020190","chr4:6020367",
                   "chr4:6025638","chr4:6025656","chr4:6025766","chr4:6026058",
                   "chr4:6083488","chr4:6204935","chr4:6235553","chr4:6237142",
                   "chr4:6238466","chr4:6239906","chr4:6240929","chr4:6245022",
                   "chr4:6245618","chr4:6245732","chr4:6245915","chr4:6246075",
                   "chr4:6246373","chr4:6246959","chr4:6290594","chr4:6292020",
                   "chr4:6294095","chr4:6298375","chr4:6316092","chr4:6321396",
                   "chr4:6324647","chr4:6324785","chr4:6327669","chr4:6328354",
                   "chr4:6328507","chr4:6333130","chr4:6333559","chr4:6333669",
                   "chr4:6335966","chr4:6435341","chr4:6435486","chr4:6435926",
                   "chr4:6437191","chr4:6437197","chr4:6457121","chr4:6457131",
                   "chr4:6457132","chr4:6568390","chr4:6570032","chr4:6570768",
                   "chr4:6596360","chr4:6613252","chr4:6613462","chr4:6620991",
                   "chr4:6624771","chr4:6626154","chr4:6641969","chr4:6642090",
                   "chr4:6644466","chr4:6644467","chr4:6644468","chr4:6647889",
                   "chr4:6648300","chr4:6662665","chr4:6663319","chr4:6663715",
                   "chr4:6674554","chr4:6678553","chr4:6678599","chr4:6690535",
                   "chr4:6698664","chr4:6698667","chr4:6698706","chr4:6720572",
                   "chr4:6911679","chr4:6985889","chr4:6987394","chr4:7002344",
                   "chr4:7004495","chr4:7004506","chr4:7005196","chr4:7005199",
                   "chr4:7024077","chr4:7024398","chr4:7029430","chr4:7031064",
                   "chr4:7044357","chr4:7044380","chr4:7048842","chr4:7052115",
                   "chr4:7055253","chr4:7064243","chr4:7067765","chr4:7073187",
                   "chr4:7074027","chr4:7677967","chr4:7701947","chr4:7702795",
                   "chr4:7703505","chr4:7703807","chr4:7704795","chr4:7704818",
                   "chr4:7709703","chr4:7712150","chr4:7714490","chr4:7733843",
                   "chr4:7735162","chr4:7735164","chr4:7736103","chr4:7736112")
```

# 5.7) Biological Association Visualization: Manhattan Plot

* Finally, we plot the Manhattanh plot graph

```{r, eval = T}
manhattan(GWAS_TABLE_Ommit, 
          highlight = SNP_HIGHLIGHT, 
          annotateTop = T, 
          annotatePval = 0.20, 
          genomewideline = -log10(0.0001))
```

# 6) References

<div id="refs"></div>

# 7) Session Info

```{r session}
sessionInfo()
```




